crates/rethnet_eth/src/remote/cacheable_method_invocation.rs

```diff
@@ -181,22 +253,254 @@ impl<'a> TryFrom<&'a MethodInvocation> for CacheableMethodInvocation<'a> {
             | MethodInvocation::EvmSetAutomine(_)
             | MethodInvocation::EvmSetIntervalMining(_)
             | MethodInvocation::EvmSetNextBlockTimestamp(_)
-            | MethodInvocation::EvmSnapshot() => Err(MethodNotCacheableError(value.clone())),
+            | MethodInvocation::EvmSnapshot() => {
+                Err(MethodNotCacheableError::MethodInvocation(value.clone()))
+            }
+        }
+    }
+}
+
+/// A block argument specification that is cacheable.
+#[derive(Clone, Debug)]
+pub(super) enum CacheableBlockSpec<'a> {
+    /// Block number
+    Number { block_number: &'a U256 },
+    /// Block hash
+    Hash {
+        block_hash: &'a B256,
+        require_canonical: Option<bool>,
+    },
+    /// "earliest" block tag
+    Earliest,
+    /// "safe" block tag
+    Safe,
+    /// "finalized" block tag
+    Finalized,
+}
+
+/// Error type for [`CacheableBlockSpec::try_from`].
+#[derive(thiserror::Error, Debug)]
+#[error("Method is not cacheable: {0:?}")]
+pub(super) struct BlockSpecNotCacheableError(Option<BlockSpec>);
```

Comment:
I made it a standalone error, because it's a type conversion error:

```rust
impl<'a> TryFrom<&'a BlockSpec> for CacheableBlockSpec<'a> {
    type Error = BlockSpecNotCacheableError;
```

and 

```rust
impl<'a> TryFrom<&'a BlockSpec> for CacheableBlockSpec<'a> {
    type Error = MethodNotCacheableError;
```

would look weird to me.

But `MethodNotCacheableError` can be constructed from `BlockSpecNotCacheableError`.